# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1460242708 14400
#      Sat Apr 09 18:58:28 2016 -0400
# Node ID cf552443c10455ede48e90c9a65ef99af6167577
# Parent  fe1447ae68cf965bd3e5f74d0220884487ded452
revise method of handling hg id for build info

* Makefile.am (HG-ID): New target and file to distribute.
* build-info.h: Don't include config.h.
* build-info.in.cc: Strip CR from line endings.
* build-info.h, build-info.in.cc
(octave_hg_id): Rename from oct_hg_id.
(octave_build_date): Rename from oct_build_date.
(octave_build_time): Rename from oct_build_time.
Change all callers.
* toplev.cc (F__octave_config_info__): Rename fields: builddate to
build_date, buildtime to build_time, hgid to hg_id, and releasedate to
* libinterp/module.mk (upate_hg_id): Delete rule.
(octinclude_HEADERS): Include build-info.h in the list.
(BUILT_SOURCES): Remove update_hg_id from the list.
(libinterp/build-info.cc): Don't depend on build-aux/mk-build-info.sh.
Depend on HG-ID instead of libinterp/hg.id.  Call sed directly here
instead of using a shell script.
* build-aux/mk-build-info-cc.in.sh: Delete.
* Makefile.am (EXTRA_DIST): Remove build-aux/mk-build-info-cc.sh.in
from the list.
(GEN_CONFIG_SHELL): Remove build-aux/mk-build-info-cc.sh
from the list.
* configure.ac (OCTAVE_CONFIG_MOVE_IF_CHANGE_FILES): Remove
build-aux/mk-build-info-cc.sh from the list.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -32,32 +32,32 @@ DOC_TARGETS =
 BUILT_DISTFILES =
 BUILT_NODISTFILES =
 EXTRA_DIST =
 
 BUILT_DISTFILES += \
   AUTHORS \
   BUGS \
   ChangeLog \
+  HG-ID \
   INSTALL.OCTAVE
 
 EXTRA_DIST += \
   CITATION \
   COPYING \
   INSTALL \
   NEWS \
   README \
   build-aux/OctJavaQry.class \
   build-aux/check-subst-vars.in.sh \
   build-aux/find-files-with-tests.sh \
   build-aux/mk-default-qt-settings.in.sh \
   build-aux/mk-f77-def.in.sh \
   build-aux/mk-mxarray-h.in.sh \
   build-aux/mk-version-h.in.sh \
-  build-aux/mk-build-info-cc.in.sh \
   build-aux/mk-octave-config-h.sh \
   build-aux/mk-opts.pl \
   build-aux/move-if-change \
   build-aux/stl_algo.h-fixed \
   build-aux/subst-config-vals.in.sh \
   build-aux/subst-cross-config-vals.in.sh \
   build-aux/subst-default-vals.in.sh \
   build-aux/subst-f77-isnan-macro.in.sh \
@@ -65,17 +65,16 @@ EXTRA_DIST += \
   run-octave.in \
   $(BUILT_DISTFILES)
 
 GEN_CONFIG_SHELL = \
   build-aux/mk-default-qt-settings.sh \
   build-aux/mk-f77-def.sh \
   build-aux/mk-mxarray-h.sh \
   build-aux/mk-version-h.sh \
-  build-aux/mk-build-info-cc.sh \
   build-aux/subst-config-vals.sh \
   build-aux/subst-cross-config-vals.sh \
   build-aux/subst-default-vals.sh \
   build-aux/subst-f77-isnan-macro.sh \
   build-aux/subst-script-vals.sh
 
 $(GEN_CONFIG_SHELL) : %.sh : %.in.sh config.status
 	$(AM_V_GEN)$(SHELL) config.status $@-tmp $@
@@ -282,16 +281,30 @@ define changelog-from-hg-log
     mv $@-t $@; \
   fi
 endef
 
 ChangeLog:
 	$(AM_V_GEN)$(changelog-from-hg-log)
 .PHONY: ChangeLog
 
+HG-ID:
+	$(AM_V_GEN)rm -f $@-t && \
+	if [ -d $(srcdir)/.hg ]; then \
+	  ( cd $(srcdir) && hg identify --id ) > $@-t && \
+	  $(simple_move_if_change_rule); \
+	elif [ ! -f $(srcdir)/HG-ID ]; then \
+	  echo "$(srcdir)/HG-ID is missing!" 1>&2; \
+	  echo "unknown" >& $@-t && mv $@-t $@; \
+	else \
+	  echo "preserving existing HG-ID file" 1>&2; \
+	  $(cp_update_rule); \
+	fi
+.PHONY: HG-ID
+
 octetc_DATA += \
   CITATION \
   NEWS
 
 if AMCOND_INSTALL_BUILD_LOGS
 octetc_DATA += config.log
 endif
 
diff --git a/build-aux/mk-build-info-cc.in.sh b/build-aux/mk-build-info-cc.in.sh
deleted file mode 100644
--- a/build-aux/mk-build-info-cc.in.sh
+++ /dev/null
@@ -1,27 +0,0 @@
-#! /bin/sh
-#
-# Copyright (C) 2016 M. Muetzel
-#
-# This file is part of Octave.
-#
-# Octave is free software; you can redistribute it and/or modify it
-# under the terms of the GNU General Public License as published by the
-# Free Software Foundation; either version 3 of the License, or (at
-# your option) any later version.
-#
-# Octave is distributed in the hope that it will be useful, but WITHOUT
-# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
-# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
-# for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with Octave; see the file COPYING.  If not, see
-# <http://www.gnu.org/licenses/>.
-
-: ${SED=@SED@}
-
-OCTAVE_HG_ID=$(hg identify --id)
-
-$SED \
-  -e "s|%NO_EDIT_WARNING%|DO NOT EDIT!  Generated automatically by mk-build-info-cc.|" \
-  -e "s|%OCTAVE_HG_ID%|${OCTAVE_HG_ID}|"
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -3394,17 +3394,16 @@ AC_CONFIG_FILES([oct-conf-post.h-tmp:oct
 AC_CONFIG_COMMANDS([oct-conf-post.h],
   [$SHELL $srcdir/build-aux/move-if-change oct-conf-post.h-tmp oct-conf-post.h])
 
 OCTAVE_CONFIG_MOVE_IF_CHANGE_FILES([
   build-aux/mk-default-qt-settings.sh
   build-aux/mk-f77-def.sh
   build-aux/mk-mxarray-h.sh
   build-aux/mk-version-h.sh
-  build-aux/mk-build-info-cc.sh
   build-aux/subst-config-vals.sh
   build-aux/subst-cross-config-vals.sh
   build-aux/subst-default-vals.sh
   build-aux/subst-f77-isnan-macro.sh
   build-aux/subst-script-vals.sh])
 
 AC_OUTPUT
 
diff --git a/libinterp/build-info.h b/libinterp/build-info.h
--- a/libinterp/build-info.h
+++ b/libinterp/build-info.h
@@ -16,19 +16,15 @@ FITNESS FOR A PARTICULAR PURPOSE.  See t
 for more details.
 
 You should have received a copy of the GNU General Public License
 along with Octave; see the file COPYING.  If not, see
 <http://www.gnu.org/licenses/>.
 
 */
 
-#ifdef HAVE_CONFIG_H
-#  include "config.h"
-#endif
-
 #include <string>
 
-OCTAVE_API std::string oct_hg_id (void);
+extern OCTAVE_API std::string octave_hg_id (void);
 
-OCTAVE_API std::string oct_build_date (void);
+extern OCTAVE_API std::string octave_build_date (void);
 
-OCTAVE_API std::string oct_build_time (void);
+extern OCTAVE_API std::string octave_build_time (void);
diff --git a/libinterp/build-info.in.cc b/libinterp/build-info.in.cc
--- a/libinterp/build-info.in.cc
+++ b/libinterp/build-info.in.cc
@@ -1,39 +1,46 @@
-// %NO_EDIT_WARNING%
-/*
-
-Copyright (C) 2016 M. Muetzel
-
-This file is part of Octave.
-
-Octave is free software; you can redistribute it and/or modify it
-under the terms of the GNU General Public License as published by the
-Free Software Foundation; either version 3 of the License, or (at your
-option) any later version.
-
-Octave is distributed in the hope that it will be useful, but WITHOUT
-ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
-FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
-for more details.
-
-You should have received a copy of the GNU General Public License
-along with Octave; see the file COPYING.  If not, see
-<http://www.gnu.org/licenses/>.
-
-*/
-
-#include "build-info.h"
-
-std::string oct_hg_id (void)
-{
-  return "%OCTAVE_HG_ID%";
-}
-
-std::string oct_build_date (void)
-{
-  return __DATE__;
-}
-
-std::string oct_build_time (void)
-{
-  return __TIME__;
-}
+// %NO_EDIT_WARNING%
+/*
+
+Copyright (C) 2016 M. Muetzel
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "build-info.h"
+
+std::string
+octave_hg_id (void)
+{
+  return "%OCTAVE_HG_ID%";
+}
+
+std::string
+octave_build_date (void)
+{
+  return __DATE__;
+}
+
+std::string
+octave_build_time (void)
+{
+  return __TIME__;
+}
diff --git a/libinterp/corefcn/toplev.cc b/libinterp/corefcn/toplev.cc
--- a/libinterp/corefcn/toplev.cc
+++ b/libinterp/corefcn/toplev.cc
@@ -1418,24 +1418,24 @@ specified option.\n\
       { "ENABLE_OPENMP", true },
 #else
       { "ENABLE_OPENMP", false },
 #endif
 
       { "api_version", OCTAVE_API_VERSION },
       { "archlibdir", subst_octave_home (OCTAVE_ARCHLIBDIR) },
       { "bindir", subst_octave_home (OCTAVE_BINDIR) },
-      { "builddate", oct_build_date () },
-      { "buildtime", oct_build_time () },
+      { "build_date", octave_build_date () },
+      { "build_time", octave_build_time () },
       { "canonical_host_type", OCTAVE_CANONICAL_HOST_TYPE },
       { "datadir", subst_octave_home (OCTAVE_DATADIR) },
       { "datarootdir", subst_octave_home (OCTAVE_DATAROOTDIR) },
       { "exec_prefix", subst_octave_home (OCTAVE_EXEC_PREFIX) },
       { "fcnfiledir", subst_octave_home (OCTAVE_FCNFILEDIR) },
-      { "hgid", oct_hg_id () },
+      { "hg_id", octave_hg_id () },
       { "imagedir", subst_octave_home (OCTAVE_IMAGEDIR) },
       { "includedir", subst_octave_home (OCTAVE_INCLUDEDIR) },
       { "infodir", subst_octave_home (OCTAVE_INFODIR) },
       { "infofile", subst_octave_home (OCTAVE_INFOFILE) },
       { "libdir", subst_octave_home (OCTAVE_LIBDIR) },
       { "libexecdir", subst_octave_home (OCTAVE_LIBEXECDIR) },
       { "localapiarchlibdir", subst_octave_home (OCTAVE_LOCALAPIARCHLIBDIR) },
       { "localapifcnfiledir", subst_octave_home (OCTAVE_LOCALAPIFCNFILEDIR) },
@@ -1452,17 +1452,17 @@ specified option.\n\
       { "mandir", subst_octave_home (OCTAVE_MANDIR) },
       { "octdatadir", subst_octave_home (OCTAVE_OCTDATADIR) },
       { "octfiledir", subst_octave_home (OCTAVE_OCTFILEDIR) },
       { "octetcdir", subst_octave_home (OCTAVE_OCTETCDIR) },
       { "octincludedir", subst_octave_home (OCTAVE_OCTINCLUDEDIR) },
       { "octlibdir", subst_octave_home (OCTAVE_OCTLIBDIR) },
       { "octtestsdir", subst_octave_home (OCTAVE_OCTTESTSDIR) },
       { "prefix", subst_octave_home (OCTAVE_PREFIX) },
-      { "releasedate", OCTAVE_RELEASE_DATE },
+      { "release_date", OCTAVE_RELEASE_DATE },
       { "startupfiledir", subst_octave_home (OCTAVE_STARTUPFILEDIR) },
       { "version", OCTAVE_VERSION },
       { 0, octave_value () }
     };
 
   struct build_info_struct
   {
     const char *key;
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -38,17 +38,16 @@ GENERATED_MAKE_BUILTINS_INCS = \
   libinterp/corefcn/mxarray.h \
   libinterp/parse-tree/oct-gperf.h \
   libinterp/parse-tree/oct-parse.h \
   libinterp/version.h
 
 BUILT_SOURCES += \
   $(GENERATED_MAKE_BUILTINS_INCS) \
   libinterp/build-env.cc \
-  update_hg_id \
   libinterp/build-info.cc \
   libinterp/builtin-defun-decls.h \
   libinterp/builtins.cc \
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/oct-tex-lexer.cc \
   libinterp/corefcn/oct-tex-parser.cc \
   libinterp/corefcn/oct-tex-symbols.cc \
   libinterp/operators/ops.cc \
@@ -106,16 +105,17 @@ libinterp_EXTRA_DIST += \
   libinterp/version.in.h \
   $(LIBINTERP_BUILT_DISTFILES)
 
 octinclude_HEADERS += \
   libinterp/corefcn/graphics-props.cc \
   libinterp/builtins.h \
   libinterp/builtin-defun-decls.h \
   libinterp/build-env.h \
+  libinterp/build-info.h \
   libinterp/octave.h \
   libinterp/options-usage.h \
   $(OCTAVE_VALUE_INC) \
   $(PARSE_TREE_INC) \
   $(PARSER_INC) \
   $(COREFCN_INC)
 
 noinst_HEADERS += \
@@ -258,24 +258,22 @@ libinterp/build-env.cc: libinterp/build-
 libinterp/build-env-features.cc: config.h libinterp/build-env-features.sh | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SHELL) $(srcdir)/libinterp/build-env-features.sh $< > $@-t && \
 	$(simple_move_if_change_rule)
 
 libinterp/version.h: libinterp/version.in.h build-aux/mk-version-h.sh | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)$(call simple-filter-rule,build-aux/mk-version-h.sh)
 
-update_hg_id:
-	@if [ "x$(shell cat libinterp/hg.id)" != "x$(shell hg identify --id)" ]; then \
-		hg identify --id > libinterp/hg.id; \
-	fi
-.PHONY: update_hg_id
-
-libinterp/build-info.cc: libinterp/build-info.in.cc build-aux/mk-build-info-cc.sh libinterp/hg.id | libinterp/$(octave-dirstamp)
-	$(AM_V_GEN)$(call simple-filter-rule,build-aux/mk-build-info.sh)
+libinterp/build-info.cc: libinterp/build-info.in.cc HG-ID | libinterp/$(octave-dirstamp)
+	$(AM_V_GEN)rm -f $@-t && \
+	$(SED) \
+	  -e "s|%NO_EDIT_WARNING%|DO NOT EDIT!  Generated automatically by Makefile|" \
+	  -e "s|%OCTAVE_HG_ID%|`cat $(builddir)/HG-ID`|" $< > $@-t && \
+	$(simple_move_if_change_rule)
 
 libinterp/builtins.cc: $(DEF_FILES) libinterp/mkbuiltins | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SHELL) $(srcdir)/libinterp/mkbuiltins --source $(DEF_FILES) > $@-t && \
 	$(simple_move_if_change_rule)
 
 libinterp/builtin-defun-decls.h: $(SRC_DEF_FILES) libinterp/mkbuiltins | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
